### ISSUE TITLE
How to use with pandas dataframe?

### ISSUE THREAD

**Issue by anushiya-thevapalan** [MilaNLProc/contextualized-topic-models#4]
I have a pandas data frame in which a column has a set of tweets. How do I process it using this library and get the output to another column in the data frame?

**Comment by vinid** [comment]
Hi!

I'm assuming you are interested in getting the topic for each tweet: I think the easiest way to do this is to 1) write the files to a text file, (2) run the model and then (3) create a new column into the pandas dataframe.

Let's assume you have a dataframe with two columns "tweet_text" and "tweet_text_preprocessed" (having preprocessed and clean text is not a requirement, but it might reduce the amount of noise. Also, it might be better to get the contextual from tweets that have not been too preprocessed, since this might help SBERT in creating good representations).

```python
with open("pre_documents.txt", "w") as filino:
    for line in df["tweet_text_preprocessed"].values:
        filino.write(line + "\n")

with open("documents.txt", "w") as filino:
    for line in df["tweet_text"].values:
        filino.write(line + "\n")

handler = TextHandler("pre_documents.txt")
handler.prepare() 

training_bert = bert_embeddings_from_file("documents.txt", "bert-base-nli-mean-tokens") # I'm assuming the tweets are in english here.

training_dataset = CTMDataset(handler.bow, training_bert, handler.idx2token)

ctm = CTM(input_size=len(handler.vocab), bert_input_size=768, 
      inference_type="combined", n_components=n_topics)

ctm.fit(training_dataset)
```

-------

After the model has been trained, you can get the topic for each document with the following code (you can also find this snippet in the multilingual notebook, albeit in a different context). Essentially, we sample the distribution 100 times (ctm.get_thetas(training_dataset)) and get a reliable estimate of the topic by normalizing (beware, this is not the best and fastest way to do this, and it will take a bit to complete)

```python
import numpy as np
predicted_topics = [] 

# create thetas: num_documents X num_topics

thetas = np.zeros((len(training_dataset), n_topics))
for a in range(0, 100):
    # sample multiple times
    thetas = thetas + np.array(ctm.get_thetas(training_dataset))
    
# normalize thetas for each document and get the topics
for idd in range(0, len(training_dataset)):
    thetas[idd] = thetas[idd]/np.sum(thetas[idd]) 
    predicted_topic = np.argmax(thetas[idd]) 
    predicted_topics.append(predicted_topic)

```
**thetas** contains the topic distribution (i.e., the probability of each topic, given a document). thetas[0] contains the distribution for the first document. **np.argmax(thetas[0])** gives you a number, that is the index of the topic with the highest probability. **predicted_topics** is a list that contains the predicted topic index for each document.

Hence, you can get the actual topic words for the first document using:

```python 
ctm.get_topic_lists()[predicted_topics[0]]
```

Eventually, you can do something like

```python
collect = []

for i in range(0, len(df)):
    collect.append(ctm.get_topic_lists()[predicted_topics[i])

df["topics"] = collect
```

This should be a good part of the code needed to get you where you wanted :)

Ping me if you need more help :)

**Comment by anushiya-thevapalan** [comment]
Thank you so much for a very good explanation with code snippets. That's good enough to get me started. I just wanted to clarify one more thing.

Say I train a model using tweets in the English language and get the topics for those tweets. Now I need to get the topics of tweets that are in Spanish (But not the translation of the English tweets), fresh tweets which the model has not seen before. Will it still be able to predict the topics?

**Comment by vinid** [comment]
Yes, it will predict **english** topics for the spanish unseen documents (see the paper for more details). However, if the context of the training document is completely different from the ones used at test time (e.g., you train on sport related tweet in english and test on politics related tweets in spanish) the topic prediction will be less accurate.

A few more notes:

You will need to use multilingual embeddings in this case

```python
training_bert = bert_embeddings_from_file("documents.txt", "distiluse-base-multilingual-cased")
````
and you must use the "contextual" model in place of the "combined" one:

```python
ctm = CTM(input_size=len(handler.vocab), bert_input_size=512, inference_type="contextual", n_components=50)
```
The multilingual notebook in the example folder should contain all that is needed to run this process. 

At test time, you can simply create the new vector embeddings for a different language and predict the topics. You'll find that to predict topic you can use the very same code I've shared with you in the first message.

```python
testing_bert_italian = bert_embeddings_from_file('unpreprocessed_docs_italian.txt', "distiluse-base-multilingual-cased")
testing_dataset_italian = CTMDataset(testing_bert_italian, testing_bert_italian, []) # not the best definition, but it's an effective workaround to get the topics for italian documents.
```

and you get the topic distributions:

```python
predicted_topics = [] 
thetas = np.zeros((len(testing_dataset_italian), num_topics))
for a in range(0, 100):
    thetas = thetas + np.array(ctm.get_thetas(testing_dataset_italian))
    
for idd in range(0, len(testing_dataset_italian)):
    thetas[idd] = thetas[idd]/np.sum(thetas[idd])
    predicted_topic = np.argmax(thetas[idd]) 
    predicted_topics.append(predicted_topic)

thetas # document-topic distribution 
predicted_topics # list of the topic predicted for each testing document
```

Hope this helps :)

**Comment by anushiya-thevapalan** [comment]
Thanks a lot for the explanation. One quicker question

If we train this model on large data consisting of a vast number of contexts, will it accurately generalize? 

For example, say you train the model with tweets relating to sports, education, politics, and many more. If I wish to get the tweet belonging to either one of the above contexts, will it correctly assign the topics, irrespective of the vast number of contexts used in the training process?

If this is so, is there any pre-trained model you built?

**Comment by vinid** [comment]
Yes, obviously there is a limit to how well this can be optimized and different **number of topics** will result in different prediction granularities. From our results on Wikipedia, it seemed to work really well when tested on unseen Wikipedia documents in different languages (note that while the documents are unseen, the _structure_ is similar, since we are talking about encyclopedic knowledge.).

We are currently planning to generate some pre-trained models with bigger corpora, but it will take a while.

**Comment by anushiya-thevapalan** [comment]
Thank you so much for taking time to explain in detail.

**Comment by vinid** [comment]
No problem :)
